onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc

```diff
@@ -1030,12 +1297,25 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<onnxruntime:
       }
 
       // Cast INT64 input to INT32 because TensorRT doesn't fully support INT64
-      for (int i = 0, end = num_binding_outputs; i < end; ++i) {
-        if (output_types[i] == ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64) {
-          cuda::Impl_Cast<int32_t, int64_t>(reinterpret_cast<int32_t*>(buffers[i + num_binding_inputs]), ort.GetTensorMutableData<int64_t>(output_tensor[i]), output_dim_sizes[i]);
-          cudaDeviceSynchronize();
-          cudaFree(buffers[i + num_binding_inputs]);
+      for (int i = 0, end = output_binding_names.size(); i < end; ++i) {
+        const std::string& output_name = output_binding_names[i];
+        size_t binding_index = trt_engine->getBindingIndex(output_name.c_str());
+        int output_type = 0;
+        const auto& iter = output_types.find(output_name);
+        if (iter != output_types.end()) {
+          output_type = iter->second;
         }
+        if (output_type == ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64) {
+          auto output_tensor_ptr = ort.GetTensorMutableData<int64_t>(output_tensor[i]);
+          if (output_tensor_ptr != nullptr) {
+            cuda::Impl_Cast<int32_t, int64_t>(reinterpret_cast<int32_t*>(buffers[binding_index]), output_tensor_ptr, output_dim_sizes[i]);
+          }
+        }
+      }
+
+      cudaDeviceSynchronize();
+      for (const auto& binding_index : binding_buffers_to_freeup) {
```

Comment:
are we leaking memory if the enqueueV2() fails above?

Reply:
For FAIL status, would the session run quit or continue to try other EPs? For the latter, we may need to free buffer when FAIL occurs

Reply:
the session doesn't get destroyed automatically.
can enqueue failure be intermittent? (can there be a success after failure?)
in any case, a user could issue Run() again, or do the fallback manually (create a new session), so 
it does seem like it could cause a leak.